File: dotnet/src/Bewerbung.Api/BewerbungService.cs
using System.Security.Cryptography;

namespace Bewerbung.Api;

/// <summary>
/// Use-Case-Schicht: reine Fachlogik, kennt weder DB noch HTTP.
/// Genau deshalb ohne Datenbank unit-testbar (Fake-Repository).
/// </summary>
public sealed class BewerbungService
{
    private readonly IBewerbungRepository _repo;

    public BewerbungService(IBewerbungRepository repo)
    {
        _repo = repo;
    }

    public async Task<EinreichenResult> EinreichenAsync(BewerbungInput input)
    {
        Validate(input);

        var email = input.Email!.Trim();

        var bewerberId = await _repo.FindBewerberIdByEmailAsync(email);
        bewerberId ??= await _repo.InsertBewerberAsync(new BewerberInput(
            input.Vorname!.Trim(),
            input.Nachname!.Trim(),
            email,
            string.IsNullOrWhiteSpace(input.Telefon) ? null : input.Telefon!.Trim()
        ));

        var vorgangsNr = GenerateVorgangsNr();
        var bewerbungId = await _repo.InsertBewerbungAsync(
            bewerberId.Value,
            input.StelleId!.Value,
            vorgangsNr,
            string.IsNullOrWhiteSpace(input.Bemerkung) ? null : input.Bemerkung!.Trim()
        );

        return new EinreichenResult(bewerbungId, bewerberId.Value, vorgangsNr);
    }

    public Task<IReadOnlyList<IDictionary<string, object?>>> ListeAsync(string? status)
        => _repo.ListBewerbungenAsync(status);

    public static string GenerateVorgangsNr()
        => $"BEW-{DateTime.UtcNow.Year}-{RandomNumberGenerator.GetInt32(0, 0x1000000):X6}";

    private static void Validate(BewerbungInput i)
    {
        var errors = new List<string>();

        if (string.IsNullOrWhiteSpace(i.Vorname))
            errors.Add("Feld 'vorname' ist ein Pflichtfeld.");
        if (string.IsNullOrWhiteSpace(i.Nachname))
            errors.Add("Feld 'nachname' ist ein Pflichtfeld.");

        var email = (i.Email ?? string.Empty).Trim();
        var at = email.IndexOf('@');
        if (at < 1 || !email[(at + 1)..].Contains('.'))
        {
            errors.Add("Feld 'email' ist keine gültige E-Mail-Adresse.");
        }

        if (i.StelleId is null || i.StelleId <= 0)
        {
            errors.Add("Feld 'stelle_id' muss eine positive Zahl sein.");
        }

        if (errors.Count > 0) throw new ValidationException(errors);
    }
}

public sealed record BewerbungInput(
    string? Vorname,
    string? Nachname,
    string? Email,
    string? Telefon,
    int? StelleId,
    string? Bemerkung
);

public sealed record EinreichenResult(int BewerbungId, int BewerberId, string VorgangsNr);

File: java/src/main/java/de/train/bewerbung/BewerbungService.java
package de.train.bewerbung;

import java.security.SecureRandom;
import java.sql.SQLException;
import java.time.Year;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Use-Case-Schicht: reine Fachlogik, kennt weder DB noch HTTP.
 * Genau deshalb ohne Datenbank unit-testbar (Fake-Repository).
 */
public final class BewerbungService {

    private static final SecureRandom RNG = new SecureRandom();

    private final BewerbungRepository repo;

    public BewerbungService(BewerbungRepository repo) {
        this.repo = repo;
    }

    public EinreichenResult einreichen(BewerbungInput input) throws SQLException {
        validate(input);

        String email = input.email().trim();

        Integer bewerberId = repo.findBewerberIdByEmail(email);
        if (bewerberId == null) {
            bewerberId = repo.insertBewerber(new BewerberInput(
                    input.vorname().trim(),
                    input.nachname().trim(),
                    email,
                    input.telefon() == null || input.telefon().isBlank()
                            ? null : input.telefon().trim()
            ));
        }

        String vorgangsNr = generateVorgangsNr();
        int bewerbungId = repo.insertBewerbung(
                bewerberId,
                input.stelleId(),
                vorgangsNr,
                input.bemerkung() == null || input.bemerkung().isBlank()
                        ? null : input.bemerkung().trim()
        );

        return new EinreichenResult(bewerbungId, bewerberId, vorgangsNr);
    }

    public List<Map<String, Object>> liste(String status) throws SQLException {
        return repo.listBewerbungen(status);
    }

    public static String generateVorgangsNr() {
        int n = RNG.nextInt(0x1000000);
        return String.format("BEW-%d-%06X", Year.now().getValue(), n);
    }

    private static void validate(BewerbungInput i) {
        List<String> errors = new ArrayList<>();

        if (i == null || i.vorname() == null || i.vorname().trim().isEmpty()) {
            errors.add("Feld 'vorname' ist ein Pflichtfeld.");
        }
        if (i == null || i.nachname() == null || i.nachname().trim().isEmpty()) {
            errors.add("Feld 'nachname' ist ein Pflichtfeld.");
        }

        String email = i == null || i.email() == null ? "" : i.email().trim();
        int at = email.indexOf('@');
        if (at < 1 || !email.substring(at + 1).contains(".")) {
            errors.add("Feld 'email' ist keine gueltige E-Mail-Adresse.");
        }

        Integer stelle = i == null ? null : i.stelleId();
        if (stelle == null || stelle <= 0) {
            errors.add("Feld 'stelle_id' muss eine positive Zahl sein.");
        }

        if (!errors.isEmpty()) {
            throw new ValidationException(errors);
        }
    }
}
